optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -235,11 +235,21 @@ private BigDecimalConstraintMatchTotal findBigDecimalConstraintMatchTotal(RuleCo
         return matchTotal;
     }
 
-    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext) {
+    private AgendaItem prepareAgendaItemForUnMatchListener(RuleContext kcontext, int scoreLevel) {
         AgendaItem agendaItem = (AgendaItem) kcontext.getMatch();
-        if (agendaItem.getActivationUnMatchListener() != null) {
-            // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
-            agendaItem.getActivationUnMatchListener().unMatch(null, null);
+        ConstraintMatchTotal matchTotal = null;
+        if (constraintMatchEnabled) {
+            Rule rule = kcontext.getRule();
+            matchTotal = constraintMatchTotalMap.get(Arrays.asList(rule.getPackageName(), rule.getName(), scoreLevel));
+            if (agendaItem.getActivationUnMatchListener() != null && matchTotal != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
+        } else {
+            if (agendaItem.getActivationUnMatchListener() != null) {
+                // Both parameters null because they are not used by the ActivationUnMatchListener created anyway
+                agendaItem.getActivationUnMatchListener().unMatch(null, null);
+            }
```

Comment:
I don't think we can actually make that AgendaItem unique for 2 addConstraintMatch calls for the same score rule. Which means we probably need to always register is a multiplex ActivationUnMatchListener to the AgendaItem and add our specific ActivationUnMatchListener in that multiplexer... The mutiplexer can have at most 1 ActivationUnMatchListener per score level.